File: src/dht.rs
//! Distributed Hash Table (DHT) implementation using Kademlia-style routing

use std::collections::{BTreeMap, HashMap};
use std::time::{Duration, Instant};
use crate::crypto::NodeIdentity;
use crate::message::{NodeInfo, Endpoint};
use crate::error::{Error, Result};
use hex;

/// Size of K-bucket (number of entries per bucket)
pub const K_BUCKET_SIZE: usize = 20;

/// Number of bits in node ID (SHA-256 = 256 bits)
pub const ID_BITS: usize = 256;

/// Default replication factor (k parameter)
pub const REPLICATION_FACTOR: usize = 20;

/// Value expiration time
pub const VALUE_TTL: Duration = Duration::from_secs(24 * 60 * 60); // 24 hours

/// Refresh interval for buckets
pub const BUCKET_REFRESH: Duration = Duration::from_secs(60 * 60); // 1 hour

/// A single entry in a K-bucket
#[derive(Debug, Clone)]
pub struct BucketEntry {
    pub node_id: [u8; 32],
    pub endpoint: Endpoint,
    pub last_seen: Instant,
    pub is_self: bool,
}

impl BucketEntry {
    pub fn new(node_id: [u8; 32], endpoint: Endpoint) -> Self {
        Self {
            node_id,
            endpoint,
            last_seen: Instant::now(),
            is_self: false,
        }
    }
    
    pub fn mark_seen(&mut self) {
        self.last_seen = Instant::now();
    }
}

/// A K-bucket containing up to K node entries
#[derive(Debug, Clone)]
pub struct KBucket {
    entries: Vec<BucketEntry>,
    max_size: usize,
}

impl KBucket {
    pub fn new(max_size: usize) -> Self {
        Self {
            entries: Vec::with_capacity(max_size),
            max_size,
        }
    }
    
    /// Add or update an entry
    pub fn add(&mut self, entry: BucketEntry) -> Option<BucketEntry> {
        // Check if entry already exists
        if let Some(existing) = self.entries.iter_mut().find(|e| e.node_id == entry.node_id) {
            existing.mark_seen();
            existing.endpoint = entry.endpoint;
            return None;
        }
        
        // If bucket is full, return the oldest entry for potential eviction
        if self.entries.len() >= self.max_size {
            return Some(entry);
        }
        
        self.entries.push(entry);
        None
    }
    
    /// Remove an entry
    pub fn remove(&mut self, node_id: &[u8; 32]) -> Option<BucketEntry> {
        if let Some(pos) = self.entries.iter().position(|e| &e.node_id == node_id) {
            Some(self.entries.remove(pos))
        } else {
            None
        }
    }
    
    /// Get all entries
    pub fn entries(&self) -> &[BucketEntry] {
        &self.entries
    }
    
    /// Get the oldest entry (for eviction)
    pub fn oldest(&self) -> Option<&BucketEntry> {
        self.entries.iter().min_by_key(|e| e.last_seen)
    }
    
    /// Check if bucket is full
    pub fn is_full(&self) -> bool {
        self.entries.len() >= self.max_size
    }
}

/// Stored value in the DHT
#[derive(Debug, Clone)]
pub struct StoredValue {
    pub key: Vec<u8>,
    pub value: Vec<u8>,
    pub publisher: [u8; 32],
    pub created_at: Instant,
    pub expires_at: Instant,
    pub signature: Vec<u8>,
}

impl StoredValue {
    pub fn new(key: Vec<u8>, value: Vec<u8>, publisher: [u8; 32], signature: Vec<u8>) -> Self {
        let now = Instant::now();
        Self {
            key,
            value,
            publisher,
            created_at: now,
            expires_at: now + VALUE_TTL,
            signature,
        }
    }
    
    pub fn is_expired(&self) -> bool {
        Instant::now() > self.expires_at
    }
}

/// Kademlia-style DHT routing table
pub struct RoutingTable {
    /// Our node ID
    node_id: [u8; 32],
    /// K-buckets organized by distance (index = position of first differing bit)
    buckets: Vec<KBucket>,
    /// Our own entry
    self_entry: BucketEntry,
}

impl RoutingTable {
    pub fn new(node_id: [u8; 32], endpoint: Endpoint) -> Self {
        let mut buckets = Vec::with_capacity(ID_BITS);
        for _ in 0..ID_BITS {
            buckets.push(KBucket::new(K_BUCKET_SIZE));
        }
        
        let mut self_entry = BucketEntry::new(node_id, endpoint);
        self_entry.is_self = true;
        
        Self {
            node_id,
            buckets,
            self_entry,
        }
    }
    
    /// Get the bucket index for a given distance
    fn bucket_index(distance: &[u8; 32]) -> usize {
        for i in 0..ID_BITS {
            let byte_index = i / 8;
            let bit_index = 7 - (i % 8);
            if distance[byte_index] & (1 << bit_index) != 0 {
                return i;
            }
        }
        0
    }
    
    /// Add a node to the routing table
    pub fn add_node(&mut self, node_id: [u8; 32], endpoint: Endpoint) -> Result<()> {
        if node_id == self.node_id {
            return Ok(()); // Don't add self
        }
        
        let distance = NodeIdentity::distance(&self.node_id, &node_id);
        let index = Self::bucket_index(&distance);
        
        let entry = BucketEntry::new(node_id, endpoint);
        
        if let Some(evict_candidate) = self.buckets[index].add(entry) {
            // Bucket is full, could implement ping-based eviction here
            // For now, just return the candidate for potential removal
            return Err(Error::Dht(format!(
                "Bucket {} full, candidate for eviction: {:?}",
                index,
                hex::encode(&evict_candidate.node_id)
            )));
        }
        
        Ok(())
    }
    
    /// Remove a node from the routing table
    pub fn remove_node(&mut self, node_id: &[u8; 32]) {
        if node_id == &self.node_id {
            return;
        }
        
        let distance = NodeIdentity::distance(&self.node_id, node_id);
        let index = Self::bucket_index(&distance);
        self.buckets[index].remove(node_id);
    }
    
    /// Find the k closest nodes to a target
    pub fn find_closest(&self, target: &[u8; 32], k: usize) -> Vec<NodeInfo> {
        let mut all_nodes: Vec<(&BucketEntry, [u8; 32])> = Vec::new();
        
        // Collect all nodes with their distances
        for bucket in &self.buckets {
            for entry in bucket.entries() {
                let distance = NodeIdentity::distance(&entry.node_id, target);
                all_nodes.push((entry, distance));
            }
        }
        
        // Sort by distance
        all_nodes.sort_by(|a, b| {
            NodeIdentity::compare_distance(&a.1, &b.1)
        });
        
        // Return top k
        all_nodes.into_iter()
            .take(k)
            .map(|(entry, distance)| NodeInfo {
                node_id: entry.node_id.to_vec(),
                endpoint: entry.endpoint.clone(),
                distance: u32::from_be_bytes([
                    distance[0], distance[1], distance[2], distance[3]
                ]),
            })
            .collect()
    }
    
    /// Get all nodes in routing table
    pub fn all_nodes(&self) -> Vec<NodeInfo> {
        let mut nodes = Vec::new();
        
        for bucket in &self.buckets {
            for entry in bucket.entries() {
                nodes.push(NodeInfo {
                    node_id: entry.node_id.to_vec(),
                    endpoint: entry.endpoint.clone(),
                    distance: 0, // Not calculated for all_nodes
                });
            }
        }
        
        nodes
    }
    
    /// Get count of nodes in routing table
    pub fn node_count(&self) -> usize {
        self.buckets.iter().map(|b| b.entries().len()).sum()
    }
    
    /// Get our node ID
    pub fn node_id(&self) -> &[u8; 32] {
        &self.node_id
    }
}

/// DHT storage for key-value pairs
pub struct DhtStorage {
    /// Stored values indexed by key hash
    values: HashMap<Vec<u8>, StoredValue>,
    /// Maximum number of values to store
    max_values: usize,
}

impl DhtStorage {
    pub fn new(max_values: usize) -> Self {
        Self {
            values: HashMap::new(),
            max_values,
        }
    }
    
    /// Store a value
    pub fn store(&mut self, value: StoredValue) -> Result<()> {
        // Check if we need to evict
        if self.values.len() >= self.max_values && !self.values.contains_key(&value.key) {
            // Evict expired values first
            let expired: Vec<_> = self.values.iter()
                .filter(|(_, v)| v.is_expired())
                .map(|(k, _)| k.clone())
                .collect();
            
            for key in expired {
                self.values.remove(&key);
            }
            
            // If still full, evict oldest
            if self.values.len() >= self.max_values {
                let oldest_key = self.values.iter()
                    .min_by_key(|(_, v)| v.created_at)
                    .map(|(k, _)| k.clone());
                
                if let Some(key) = oldest_key {
                    self.values.remove(&key);
                }
            }
        }
        
        self.values.insert(value.key.clone(), value);
        Ok(())
    }
    
    /// Get a value by key
    pub fn get(&self, key: &[u8]) -> Option<&StoredValue> {
        self.values.get(key).filter(|v| !v.is_expired())
    }
    
    /// Remove expired values
    pub fn cleanup(&mut self) -> usize {
        let expired: Vec<_> = self.values.iter()
            .filter(|(_, v)| v.is_expired())
            .map(|(k, _)| k.clone())
            .collect();
        
        let count = expired.len();
        for key in expired {
            self.values.remove(&key);
        }
        count
    }
    
    /// Get count of stored values
    pub fn value_count(&self) -> usize {
        self.values.len()
    }
}

/// Main DHT struct combining routing and storage
pub struct Dht {
    routing_table: RoutingTable,
    storage: DhtStorage,
    /// Pending queries (for iterative lookups)
    pending_queries: HashMap<Vec<u8>, Instant>,
    /// Query timeout
    query_timeout: Duration,
}

impl Dht {
    pub fn new(node_id: [u8; 32], endpoint: Endpoint, max_values: usize) -> Self {
        Self {
            routing_table: RoutingTable::new(node_id, endpoint),
            storage: DhtStorage::new(max_values),
            pending_queries: HashMap::new(),
            query_timeout: Duration::from_secs(30),
        }
    }
    
    /// Get the routing table
    pub fn routing_table(&self) -> &RoutingTable {
        &self.routing_table
    }
    
    /// Get mutable routing table
    pub fn routing_table_mut(&mut self) -> &mut RoutingTable {
        &mut self.routing_table
    }
    
    /// Get storage
    pub fn storage(&self) -> &DhtStorage {
        &self.storage
    }
    
    /// Get mutable storage
    pub fn storage_mut(&mut self) -> &mut DhtStorage {
        &mut self.storage
    }
    
    /// Store a value locally
    pub fn store(&mut self, key: Vec<u8>, value: Vec<u8>, publisher: [u8; 32], signature: Vec<u8>) -> Result<()> {
        let stored_value = StoredValue::new(key.clone(), value, publisher, signature);
        self.storage.store(stored_value)
    }
    
    /// Get a value locally
    pub fn get(&self, key: &[u8]) -> Option<&StoredValue> {
        self.storage.get(key)
    }
    
    /// Find closest nodes to a target
    pub fn find_closest_nodes(&self, target: &[u8; 32], k: usize) -> Vec<NodeInfo> {
        self.routing_table.find_closest(target, k)
    }
    
    /// Add a node to routing table
    pub fn add_node(&mut self, node_id: [u8; 32], endpoint: Endpoint) -> Result<()> {
        self.routing_table.add_node(node_id, endpoint)
    }
    
    /// Remove a node from routing table
    pub fn remove_node(&mut self, node_id: &[u8; 32]) {
        self.routing_table.remove_node(node_id);
    }
    
    /// Start a query (track as pending)
    pub fn start_query(&mut self, key: Vec<u8>) {
        self.pending_queries.insert(key, Instant::now());
    }
    
    /// Complete a query
    pub fn complete_query(&mut self, key: &[u8]) {
        self.pending_queries.remove(key);
    }
    
    /// Cleanup timed out queries
    pub fn cleanup_queries(&mut self) -> Vec<Vec<u8>> {
        let now = Instant::now();
        let timed_out: Vec<_> = self.pending_queries.iter()
            .filter(|(_, started)| now.duration_since(**started) > self.query_timeout)
            .map(|(k, _)| k.clone())
            .collect();
        
        for key in &timed_out {
            self.pending_queries.remove(key);
        }
        
        timed_out
    }
    
    /// Check if we are responsible for a key (closest to key among known nodes)
    pub fn is_responsible_for(&self, key: &[u8]) -> bool {
        let closest = self.routing_table.find_closest(key, 1);
        
        if closest.is_empty() {
            return true; // We're the only node
        }
        
        // Compare our distance to the closest known node's distance
        let our_distance = NodeIdentity::distance(self.routing_table.node_id(), &{
            let mut arr = [0u8; 32];
            arr.copy_from_slice(&key[..32.min(key.len())]);
            arr
        });
        
        let their_distance = {
            let mut arr = [0u8; 32];
            arr.copy_from_slice(&closest[0].node_id);
            NodeIdentity::distance(&arr, &{
                let mut arr2 = [0u8; 32];
                arr2.copy_from_slice(&key[..32.min(key.len())]);
                arr2
            })
        };
        
        NodeIdentity::compare_distance(&our_distance, &their_distance) == std::cmp::Ordering::Less
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_routing_table_add_nodes() {
        let node_id = [1u8; 32];
        let endpoint = Endpoint::default();
        let mut table = RoutingTable::new(node_id, endpoint);
        
        // Add some nodes
        for i in 0..50 {
            let mut new_id = [0u8; 32];
            new_id[0] = i as u8;
            let ep = Endpoint::default();
            let _ = table.add_node(new_id, ep);
        }
        
        assert!(table.node_count() > 0);
        assert!(table.node_count() <= K_BUCKET_SIZE * ID_BITS);
    }
    
    #[test]
    fn test_routing_table_find_closest() {
        let node_id = [1u8; 32];
        let endpoint = Endpoint::default();
        let mut table = RoutingTable::new(node_id, endpoint);
        
        // Add nodes
        for i in 0..100 {
            let mut new_id = [0u8; 32];
            new_id[0] = i as u8;
            let ep = Endpoint::default();
            let _ = table.add_node(new_id, ep);
        }
        
        // Find closest to a target
        let target = [50u8; 32];
        let closest = table.find_closest(&target, 20);
        
        assert_eq!(closest.len(), 20.min(table.node_count()));
    }
    
    #[test]
    fn test_dht_store_get() {
        let node_id = [1u8; 32];
        let endpoint = Endpoint::default();
        let mut dht = Dht::new(node_id, endpoint, 1000);
        
        let key = b"test_key".to_vec();
        let value = b"test_value".to_vec();
        let publisher = [2u8; 32];
        let signature = vec![0u8; 64];
        
        dht.store(key.clone(), value.clone(), publisher, signature).unwrap();
        
        let stored = dht.get(&key);
        assert!(stored.is_some());
        assert_eq!(stored.unwrap().value, value);
    }
}

File: src/dht.cpp
/**
 * Distributed Hash Table (DHT) Implementation - C++
 */

#include "dht.hpp"
#include "crypto.hpp"
#include <algorithm>
#include <numeric>

namespace aether {

// StoredValue implementation

StoredValue::StoredValue(std::vector<uint8_t> key,
                         std::vector<uint8_t> value,
                         const NodeId& publisher,
                         const Signature& signature,
                         std::chrono::hours ttl)
    : key(std::move(key))
    , value(std::move(value))
    , publisher(publisher)
    , signature(signature)
    , created_at(std::chrono::steady_clock::now())
    , expires_at(created_at + ttl)
{}

bool StoredValue::is_expired() const {
    return std::chrono::steady_clock::now() > expires_at;
}

// BucketEntry implementation

BucketEntry::BucketEntry(const NodeId& node_id, const Endpoint& endpoint)
    : node_id(node_id)
    , endpoint(endpoint)
    , last_seen(std::chrono::steady_clock::now())
    , is_self(false)
{}

void BucketEntry::mark_seen() {
    last_seen = std::chrono::steady_clock::now();
}

// KBucket implementation

KBucket::KBucket(size_t max_size)
    : max_size_(max_size)
{}

std::optional<BucketEntry> KBucket::add(const BucketEntry& entry) {
    // Check if entry already exists
    auto it = std::find_if(entries_.begin(), entries_.end(),
        [&entry](const BucketEntry& e) {
            return e.node_id == entry.node_id;
        });

    if (it != entries_.end()) {
        it->mark_seen();
        it->endpoint = entry.endpoint;
        return std::nullopt;
    }

    // If bucket is full, return the oldest entry
    if (entries_.size() >= max_size_) {
        const BucketEntry* oldest = oldest_ptr();
        if (oldest) {
            return *oldest;
        }
    }

    entries_.push_back(entry);
    return std::nullopt;
}

std::optional<BucketEntry> KBucket::remove(const NodeId& node_id) {
    auto it = std::find_if(entries_.begin(), entries_.end(),
        [&node_id](const BucketEntry& e) {
            return e.node_id == node_id;
        });

    if (it != entries_.end()) {
        BucketEntry entry = *it;
        entries_.erase(it);
        return entry;
    }

    return std::nullopt;
}

const BucketEntry* KBucket::oldest_ptr() const {
    if (entries_.empty()) {
        return nullptr;
    }

    auto it = std::min_element(entries_.begin(), entries_.end(),
        [](const BucketEntry& a, const BucketEntry& b) {
            return a.last_seen < b.last_seen;
        });

    return &(*it);
}

// RoutingTable implementation

RoutingTable::RoutingTable(const NodeId& node_id, const Endpoint& endpoint)
    : node_id_(node_id)
    , self_entry_(node_id, endpoint)
{
    self_entry_.is_self = true;
}

size_t RoutingTable::bucket_index(const std::array<uint8_t, NODE_ID_SIZE>& distance) {
    for (size_t i = 0; i < ID_BITS; i++) {
        size_t byte_idx = i / 8;
        size_t bit_idx = 7 - (i % 8);
        if (distance[byte_idx] & (1 << bit_idx)) {
            return i;
        }
    }
    return 0;
}

Error RoutingTable::add_node(const NodeId& node_id, const Endpoint& endpoint) {
    // Don't add self
    if (node_id == node_id_) {
        return Error::Ok;
    }

    auto distance = Crypto::distance(node_id_, node_id);
    size_t index = bucket_index(distance);

    BucketEntry entry(node_id, endpoint);

    if (auto evict = buckets_[index].add(entry)) {
        // Bucket is full
        return Error::NoMemory;
    }

    return Error::Ok;
}

void RoutingTable::remove_node(const NodeId& node_id) {
    if (node_id == node_id_) {
        return;
    }

    auto distance = Crypto::distance(node_id_, node_id);
    size_t index = bucket_index(distance);
    buckets_[index].remove(node_id);
}

std::vector<BucketEntry> RoutingTable::find_closest(const NodeId& target, size_t k) const {
    // Collect all nodes with distances
    std::vector<std::pair<const BucketEntry*, std::array<uint8_t, NODE_ID_SIZE>>> all_nodes;

    for (const auto& bucket : buckets_) {
        for (const auto& entry : bucket.entries()) {
            auto distance = Crypto::distance(entry.node_id, target);
            all_nodes.emplace_back(&entry, distance);
        }
    }

    // Sort by distance
    std::sort(all_nodes.begin(), all_nodes.end(),
        [](const auto& a, const auto& b) {
            return Crypto::compare_distance(a.second, b.second) < 0;
        });

    // Return top k
    std::vector<BucketEntry> result;
    result.reserve(std::min(k, all_nodes.size()));

    for (size_t i = 0; i < k && i < all_nodes.size(); i++) {
        result.push_back(*all_nodes[i].first);
    }

    return result;
}

std::vector<BucketEntry> RoutingTable::all_nodes() const {
    std::vector<BucketEntry> nodes;

    for (const auto& bucket : buckets_) {
        const auto& entries = bucket.entries();
        nodes.insert(nodes.end(), entries.begin(), entries.end());
    }

    return nodes;
}

size_t RoutingTable::node_count() const {
    size_t count = 0;
    for (const auto& bucket : buckets_) {
        count += bucket.size();
    }
    return count;
}

// DhtStorage implementation

DhtStorage::DhtStorage(size_t max_values)
    : max_values_(max_values)
{}

Error DhtStorage::store(StoredValue value) {
    std::string key_str(value.key.begin(), value.key.end());

    // Check if we need to evict
    if (values_.size() >= max_values_ && values_.find(key_str) == values_.end()) {
        // Evict expired values first
        std::vector<std::string> expired_keys;
        for (const auto& [key, val] : values_) {
            if (val.is_expired()) {
                expired_keys.push_back(key);
            }
        }

        for (const auto& key : expired_keys) {
            values_.erase(key);
        }

        // If still full, evict oldest
        if (values_.size() >= max_values_) {
            auto oldest_it = std::min_element(values_.begin(), values_.end(),
                [](const auto& a, const auto& b) {
                    return a.second.created_at < b.second.created_at;
                });

            if (oldest_it != values_.end()) {
                values_.erase(oldest_it);
            }
        }
    }

    values_[key_str] = std::move(value);
    return Error::Ok;
}

std::optional<StoredValue> DhtStorage::get(const std::vector<uint8_t>& key) const {
    std::string key_str(key.begin(), key.end());

    auto it = values_.find(key_str);
    if (it != values_.end() && !it->second.is_expired()) {
        return it->second;
    }

    return std::nullopt;
}

size_t DhtStorage::cleanup() {
    std::vector<std::string> expired_keys;

    for (const auto& [key, val] : values_) {
        if (val.is_expired()) {
            expired_keys.push_back(key);
        }
    }

    for (const auto& key : expired_keys) {
        values_.erase(key);
    }

    return expired_keys.size();
}

// Dht implementation

Dht::Dht(const NodeId& node_id, const Endpoint& endpoint, size_t max_values)
    : routing_table_(node_id, endpoint)
    , storage_(max_values)
{}

Error Dht::store(const std::vector<uint8_t>& key,
                 const std::vector<uint8_t>& value,
                 const NodeId& publisher,
                 const Signature& signature) {
    StoredValue stored(key, value, publisher, signature);
    return storage_.store(std::move(stored));
}

std::optional<std::vector<uint8_t>> Dht::get(const std::vector<uint8_t>& key) const {
    auto stored = storage_.get(key);
    if (stored) {
        return stored->value;
    }
    return std::nullopt;
}

std::vector<BucketEntry> Dht::find_closest_nodes(const NodeId& target, size_t k) const {
    return routing_table_.find_closest(target, k);
}

Error Dht::add_node(const NodeId& node_id, const Endpoint& endpoint) {
    return routing_table_.add_node(node_id, endpoint);
}

void Dht::remove_node(const NodeId& node_id) {
    routing_table_.remove_node(node_id);
}

size_t Dht::cleanup() {
    return storage_.cleanup();
}

}  // namespace aether
